geoportal/c2cgeoportal_geoportal/views/layers.py

```diff
@@ -66,10 +67,123 @@
 
 if TYPE_CHECKING:
     from c2cgeoportal_commons.models import main  # pylint: disable=ungrouped-imports.useless-suppression
+
+
 LOG = logging.getLogger(__name__)
 CACHE_REGION = get_region("std")
 
 
+class _LogLastUpdate:
+    def __init__(self, layer: "main.Layer"):
+        self.layer = layer
+
+    def update(self, request: pyramid.request.Request, obj: Any) -> None:
+        last_update_date = Layers.get_metadata(self.layer, "lastUpdateDateColumn")
+        if last_update_date is not None:
+            setattr(obj, last_update_date, datetime.now())
+
+        last_update_user = Layers.get_metadata(self.layer, "lastUpdateUserColumn")
+        if last_update_user is not None:
+            setattr(obj, last_update_user, request.user.id)
+
+
+def _get_base_query(request: pyramid.request.Request, layer: "main.Layer") -> sqlalchemy.orm.query.Query:
+    from c2cgeoportal_commons.models.main import (  # pylint: disable=import-outside-toplevel
+        Layer,
+        RestrictionArea,
+        Role,
+    )
+
+    allowed = models.DBSession.query(func.count(RestrictionArea.id))
+    allowed = allowed.join(RestrictionArea.roles)
+    allowed = allowed.join(RestrictionArea.layers)
+    allowed = allowed.filter(RestrictionArea.readwrite.is_(True))
+    allowed = allowed.filter(Role.id.in_(get_roles_id(request)))
+    allowed = allowed.filter(Layer.id == layer.id)
+    return allowed
+
+
+class _CreateCallback(_LogLastUpdate):
+    def __call__(self, request: pyramid.request.Request, feature: Feature, obj: Any) -> None:
+        if obj is not None:
+            raise HTTPBadRequest("Feature id should not be provided")
+        del obj
+
+        from c2cgeoportal_commons.models.main import (  # pylint: disable=import-outside-toplevel
+            RestrictionArea,
+        )
+
+        assert models.DBSession is not None
+
+        geom = feature.geometry
+        if geom and not isinstance(geom, geojson.geometry.Default):
+            shape = shapely.geometry.shape(geom)
+            srid = Layers._get_geom_col_info(self.layer)[1]
+            spatial_elt = from_shape(shape, srid=srid)
+            allowed = _get_base_query(request, self.layer)
+            allowed = allowed.filter(
+                or_(RestrictionArea.area.is_(None), RestrictionArea.area.ST_Contains(spatial_elt))
+            )
+            if allowed.scalar() == 0:
+                raise HTTPForbidden()
+
+            # Check if geometry is valid
+            if Layers._get_validation_setting(self.layer, request):
+                Layers._validate_geometry(spatial_elt)
+
+
+class _UpdateCallback(_LogLastUpdate):
```

Comment:
I would have used composition instead of inheritance here.

Reply:
After renaming _LogLastUpdate to _BaseCallback I think that's clearer to use inheritance (event if composition is possible).